### PR TITLE
[FLINK-30014][hotfix] Ensure saving non-null value in aggregate util function state

### PR DESCRIPTION
**Pull request by yunfengzhou-hub**
## What is the purpose of the change

This PR fixes the NPE that would possibly be thrown when using `DataStreamUtils.aggregate`.

## Brief change log

  - Save only non-null values in the aggregate utility function's state.
  - Clarify the behavior of the aggregate utility function in respect of input parallelism and add a test for the corner case.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-ml-core/src/test/java/org/apache/flink/ml/common/datastream/DataStreamUtilsTest.java [hunk 1]
```diff
@@ -86,6 +87,29 @@ public void testAggregate() throws Exception {
...
+        DataStream<Long> dataStream =
+                env.fromParallelCollection(new NumberSequenceIterator(0L, 19L), Types.LONG);
+        DataStream<String> result =
+                DataStreamUtils.aggregate(
+                        dataStream, new TestAggregateFuncWithNonNeutralInitialAccumulator());
+        List<String> stringSumList = IteratorUtils.toList(result.executeAndCollect());
+        assertEquals(1, stringSumList.size());
+        String stringSum1 = stringSumList.get(0);
+
+        env.setParallelism(env.getParallelism() + 1);
```
Comment (zhipeng93) [inline flink-ml-core/src/test/java/org/apache/flink/ml/common/datastream/DataStreamUtilsTest.java:101]: Let's reset the parallelism as the default value at the end of this test.

File: flink-ml-core/src/test/java/org/apache/flink/ml/common/datastream/DataStreamUtilsTest.java [hunk 2]
```diff
@@ -86,6 +87,29 @@ public void testAggregate() throws Exception {
...
+        dataStream = env.fromParallelCollection(new NumberSequenceIterator(0L, 19L), Types.LONG);
+        result =
+                DataStreamUtils.aggregate(
+                        dataStream, new TestAggregateFuncWithNonNeutralInitialAccumulator());
+        stringSumList = IteratorUtils.toList(result.executeAndCollect());
+        assertEquals(1, stringSumList.size());
+        String stringSum2 = stringSumList.get(0);
+
+        assertNotEquals(stringSum1, stringSum2);
```
Comment (zhipeng93) [inline flink-ml-core/src/test/java/org/apache/flink/ml/common/datastream/DataStreamUtilsTest.java:110]: Let's directly check the value of the output.

**Review by zhipeng93 — COMMENTED** [review]
Thanks for the fix. Left two minor comments below.

File: flink-ml-core/src/test/java/org/apache/flink/ml/common/datastream/DataStreamUtilsTest.java (hunk 1 above)
Comment (yunfengzhou-hub) [reply flink-ml-core/src/test/java/org/apache/flink/ml/common/datastream/DataStreamUtilsTest.java:101]: The parallelism of the environment would be reset in the `before()` method, which happens before each test case, so there is no need to reset the parallelism to the default value here.

**Comment by zhipeng93** [comment]
Thanks for the update. LGTM.
